### ISSUE TITLE
tf.nn.softmax makes "GPU sync failed Error" for large size inputs.

### ISSUE THREAD

**Issue by cesc-park** [tensorflow/tensorflow#4425]
### Environment info
Operating System:
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.3 LTS
Release:        14.04

### GPU info
GPU :  GTX TITAN X 
Driver :  Driver Version: 367.44 

Installed version of CUDA and cuDNN: 

CUDA 7.5
CUDNN 5.1.3

If installed from binary pip package, provide:

https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=gpu-linux/224/artifact/pip_test/whl/tensorflow-0.10.0rc0-cp27-none-linux_x86_64.whl

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

This code makes errors.

```
import tensorflow as tf
a=tf.constant(1.0,shape=[100*30, 30000]) 
b=tf.nn.softmax(a)
sess=tf.InteractiveSession()
b.eval()
sess.close()
```

This code doesn't make errors.

```
import tensorflow as tf
a=tf.constant(1.0,shape=[20*30, 30000]) 
b=tf.nn.softmax(a)
sess=tf.InteractiveSession()
b.eval()
sess.close()
```
I think this caused by input size of the softmax function.

But, I don't understand why.

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).

```
E tensorflow/stream_executor/cuda/cuda_driver.cc:1140] could not synchronize on CUDA context: CUDA_ERROR_ILLEGAL_ADDRESS :: No stack trace available
E tensorflow/stream_executor/cuda/cuda_event.cc:49] Error polling for event status: failed to query event: CUDA_ERROR_ILLEGAL_ADDRESS
F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:198] Unexpected Event status: 1
F tensorflow/core/common_runtime/gpu/gpu_util.cc:370] GPU sync failed
[1]    57277 abort (core dumped)  CUDA_VISIBLE_DEVICES=0 ipython
```

**Comment by ibab** [comment]
I've also encountered this problem using v0.10.0 in the following issue: https://github.com/ibab/tensorflow-wavenet/issues/13

Here is another example of crashing code:
```python
import numpy as np
import tensorflow as tf

x = np.ones((256, 256), dtype=np.float32)
s = tf.Session()
X = tf.placeholder(tf.float32)
y = tf.nn.softmax(X)

s.run(y, feed_dict={X: x})
```
The size of 256 seems to be significant, as it starts failing when the input dimensions go above that.

This might be a problem with the efficient reduction code in Eigen, as `cuda-memcheck` points to that part of the code, and the problem doesn't happen when using `tf.float64`.

**Comment by alquraishi** [comment]
Can confirm this problem as well running Ubuntu 14.04 and GTX Titan X with CUDA 7.5 and cuDNN 4.0.7. Although, for me it fails with anything larger than 128 x 128, e.g. 129 x 129 makes it fail.

**Comment by alexgkendall** [comment]
I've encountered this too - in the mean time a work around is to temporarily cast the tensor to float64 like @ibab suggested

**Comment by alquraishi** [comment]
For some reason I'm getting an error when I try to cast it to float64, specifically this:

    Tensor conversion requested dtype float64 for Tensor with dtype float32

Are there any known issues with upcasting to float64?

**Comment by raingo** [comment]
This is reproduced on K80 and GTX Titan.

Another workaround is to implement softmax with primitives:

https://gist.github.com/raingo/a5808fe356b8da031837

**Comment by yosinski** [comment]
Reproduced on new Titan X (Pascal) and GTX 1080. Casting suggested by @alexgkendall works:

    import numpy as np
    import tensorflow as tf
    
    x = np.ones((256, 256), dtype=np.float32)
    s = tf.Session()
    X = tf.placeholder(tf.float32)
    #y = tf.nn.softmax(X)                                                 # fails
    y = tf.cast(tf.nn.softmax(tf.cast(X, tf.float64)), tf.float32)        # works
    
    out = s.run(y, feed_dict={X: x})

**Comment by benoitsteiner** [comment]
I believe this is fixed. I'm pushing a regression test that covers the input dimensions that were previously reported as being troublesome (ie 256x256 and 129x129). I'll close this issue in a few days once the test is available unless I hear that this is still causing problems for people.

**Comment by benoitsteiner** [comment]
Regression test submitted in https://github.com/tensorflow/tensorflow/commit/9bff346137aa1c627f2dccb77fdda10863456b45

**Comment by siddhi47** [comment]
I had the same problem when i was doing some experimental stuffs on tensorflow in jupyter notebook. The problem seemed to resolve itself when I closed other kernels that were using tensorflow.
I hope this helps
